### PR TITLE
PageToolbar: Extract navbar styles and layout into a modern emotion based component

### PR DESCRIPTION
**Pull request by torkelo**
This component aims to replace all the top toolbar / navbars we have (Dashboard, Explore, Panel Edit & Alerting). It's a bit tricky to support all use cases, for example, the dashboard has a clickable page title that can be composed of parent (folder) and dashboard title, not something Explore has. Panel Edit has a back button etc. But opted to handle all uses cases in the component to more easily share the styles and alignment but can be refactored out later when the full picture emerges. 

This only updates the dashboard toolbar, leaving the other views for follow-up PRs.  The aim here is to improve the responsive behavior for the dashboard toolbar and have buttons wrap instead of just hiding them. 

TODO:
* [x] Basic PageToolbar, support pageIcon and title, and right side action buttons via children
* [x] Add support for clickable title & parent
* [x] Add support for left side actions (share & favorite etc) 
* [x] Add support for responsive layouts where buttons wrap
* [x] Folder name hides on smaller screens
* [x] Fix text ellipsis on long titles & folder names
* [x] Hide some actions in playlist mode
* [x] Support view panel mode (back button) 
* [x] Restore support fo tv & kiosk modes
* [x] Fix mobile view when the menu is absoluty position in top left corner 
* [ ] Update Enterprise and test their buttons and precense avatars

Story so far:
Screenshot from 2021-01-25 09-16-46

### CODE REVIEW

File: packages/grafana-ui/src/components/PageLayout/PageToolbar.tsx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+                size="xxl"
+                surface="dashboard"
+                aria-label={selectors.components.BackButton.backArrow}
+                onClick={onGoBack}
+              />
+            </div>
+          )}
+          <div className={styles.titleWrapper}>
+            {parent && onClickParent && (
+              <a onClick={onClickParent} className={cx(styles.titleLink, styles.parentLink)}>
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/PageLayout/PageToolbar.tsx:78]: This should be either a link with ref or a button with onClick.

File: packages/grafana-ui/src/components/PageLayout/PageToolbar.tsx [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+                {parent} <span className={styles.parentIcon}>/</span>
+              </a>
+            )}
+            {onClickTitle && (
+              <a onClick={onClickTitle} className={styles.titleLink}>
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/PageLayout/PageToolbar.tsx:83]: Same as above.

File: packages/grafana-ui/src/components/PageLayout/PageToolbar.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+                {title}
+              </a>
+            )}
+            {!onClickTitle && <div className={styles.titleText}>{title}</div>}
+          </div>
+          {leftItems &&
+            leftItems.map((child, index) => (
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/PageLayout/PageToolbar.tsx:90]:
```suggestion
            {leftItems?.map((child, index) => (
```

File: public/app/features/dashboard/components/DashNav/DashNav.tsx [hunk 4]
```diff
@@ -172,74 +182,42 @@ class DashNav extends PureComponent<Props> {
...
-          classSuffix="save"
-          tooltip="Add panel"
-          icon="panel-add"
-          onClick={onAddPanel}
-          iconType="mono"
-          iconSize="xl"
-          key="button-panel-add"
-        />
-      );
+    const tvButton = <ToolbarButton tooltip="Cycle view mode" icon="monitor" onClick={this.onToggleTVMode} />;
```
Comment (Clarity-89) [inline public/app/features/dashboard/components/DashNav/DashNav.tsx:201]: I think this also needs `key`.

File: public/app/features/dashboard/components/DashNav/DashNav.tsx [hunk 5]
```diff
@@ -172,74 +182,42 @@ class DashNav extends PureComponent<Props> {
...
+    const timeControls = (
+      <DashNavTimeControls dashboard={dashboard} location={location} onChangeTimeZone={updateTimeZoneForSession} />
```
Comment (Clarity-89) [inline public/app/features/dashboard/components/DashNav/DashNav.tsx:203]: This too is missing `key`.

**Review by Clarity-89 — APPROVED** [review]
Great job! A few minor comments. 

Also not sure if relevant for this PR but noticed two issues:
1. `ButtonSelect` seems to have an unused prop `maxMenuHeight`, which ends up being passed to `button` causing a warning, as that's not a valid HTML attribute. 
2. The selected value in the time picker looks weird on slightly narrower screens:
Screenshot 2021-01-26 at 13 45 09

File: public/app/features/dashboard/components/DashNav/DashNav.tsx (hunk 4 above)
Comment (torkelo) [reply public/app/features/dashboard/components/DashNav/DashNav.tsx:201]: ah, good catch!

File: packages/grafana-ui/src/components/PageLayout/PageToolbar.tsx (hunk 1 above)
Comment (torkelo) [reply packages/grafana-ui/src/components/PageLayout/PageToolbar.tsx:78]: yes, was copied from DashNav this way but your right this should be a button

File: packages/grafana-ui/src/components/Button/ToolbarButton.tsx [hunk 6]
```diff
@@ -31,7 +32,20 @@ export type ToolbarButtonVariant = 'default' | 'primary' | 'destructive' | 'acti
...
+      icon,
+      className,
+      children,
+      imgSrc,
+      fullWidth,
+      isOpen,
+      narrow,
+      variant = 'default',
+      iconOnly,
+      'aria-label': ariaLabel,
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/Button/ToolbarButton.tsx:46]: Nice trick, I'll try to remember this one.

File: public/app/features/explore/LiveTailButton.tsx [hunk 7]
```diff
@@ -97,21 +97,19 @@ export function LiveTailButton(props: LiveTailButtonProps) {
...
+      <Tooltip content={isLive ? <>Pause the live stream</> : <>Live stream your logs</>} placement="bottom">
         <ToolbarButton
           iconOnly={splitted}
           variant={buttonVariant}
-          icon={!isLive || isPaused ? 'play' : 'pause'}
+          icon={!isLive ? 'play' : 'pause'}
           onClick={onClickMain}
         >
-          {isLive && isPaused ? 'Paused' : 'Live'}
+          {isPaused ? 'Live' : 'Paused'}
```
Comment (hugohaggmark) [inline public/app/features/explore/LiveTailButton.tsx:112]: Should this be the other way around or maybe this is the call to action?
```suggestion
          {isPaused ? 'Paused' : 'Live'}
```

**Review by hugohaggmark — APPROVED** [review]
Super monumental work on this 🎉   I've tested some different dashboards in several responsive sizes. The only thing I found was the stop button for playlists that seems a bit different from before:

image

File: public/app/features/explore/LiveTailButton.tsx (hunk 7 above)
Comment (torkelo) [reply public/app/features/explore/LiveTailButton.tsx:112]: ah, good catch. Wrong merge when rebasing this, fixed this in another PR.

**Comment by oddlittlebird** [comment]
@grafana/docs-squad, this might affect some of our dashboard documentation and docs for dashboard features such as variables.

**Pull request merged**